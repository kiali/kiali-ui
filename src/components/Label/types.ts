export interface Labels {
  name: string;
  value: string;
}
