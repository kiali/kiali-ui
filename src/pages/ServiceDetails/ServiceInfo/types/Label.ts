export default class Label {
  key: string;
  value: string;

  constructor(key: string, value: string) {
    this.key = key;
    this.value = value;
  }

  isValid() {
    return typeof this.key === 'string' && typeof this.value === 'string';
  }
}
