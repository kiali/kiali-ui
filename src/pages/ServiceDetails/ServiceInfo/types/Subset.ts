import { TrafficPolicy } from '../../../../types/IstioObjects';
import Label from './Label';

export default class Subset {
  name: string;
  labels: { [key: string]: string };
  trafficPolicy: TrafficPolicy;

  constructor(name: string, labels: { [key: string]: string }, trafficPolicy: TrafficPolicy) {
    this.name = name;
    this.labels = labels;
    this.trafficPolicy = trafficPolicy;
  }

  public isValid() {
    return this.hasValidName() && this.hasValidLabels() && this.hasValidTrafficPolicy();
  }

  hasValidName() {
    return typeof this.name === 'string';
  }

  hasValidLabels() {
    const valid = Object.keys(this.labels).every((k, i) => new Label(k, this.labels[k]).isValid());
    return this.labels instanceof Object && valid;
  }

  hasValidTrafficPolicy() {
    return true;
  }
}
