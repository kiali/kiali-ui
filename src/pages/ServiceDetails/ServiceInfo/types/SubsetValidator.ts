import { Subset } from '../../../../types/IstioObjects';

export default class SubsetValidator {
  subset: Subset;

  constructor(subset: Subset) {
    this.subset = subset;
  }

  public isValid() {
    return this.hasValidName() && this.hasValidLabels();
  }

  hasValidName() {
    return this.hasStringType(this.subset.name);
  }

  hasValidLabels() {
    const valid = Object.keys(this.subset.labels).every((k, _i) => this.hasValidLabel(k, this.subset.labels[k]));
    return this.subset.labels instanceof Object && valid;
  }

  hasValidLabel(name: string, value: string): boolean {
    return this.hasStringType(name) && this.hasStringType(value);
  }

  hasStringType(value: any) {
    return typeof value === 'string';
  }
}
