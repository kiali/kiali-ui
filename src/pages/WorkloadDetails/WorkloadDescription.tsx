import * as React from 'react';
import { Workload } from '../../types/Workload';
import { Card, CardBody, CardHeader, Title, Tooltip, TooltipPosition } from '@patternfly/react-core';
import DetailDescription from '../../components/Details/DetailDescription';
import { style } from 'typestyle';
import Labels from '../../components/Label/Labels';
import LocalTime from '../../components/Time/LocalTime';
import { TextOrLink } from '../../components/TextOrLink';
import { renderAPILogo, renderRuntimeLogo } from '../../components/Logo/Logos';
import * as H from '../../types/Health';
import { KialiIcon } from '../../config/KialiIcon';
import { HealthIndicator } from '../../components/Health/HealthIndicator';
import { serverConfig } from '../../config';
import MissingSidecar from '../../components/MissingSidecar/MissingSidecar';
import { PFBadge, PFBadges } from '../../components/Pf/PfBadges';
import MissingLabel from '../../components/MissingLabel/MissingLabel';

type WorkloadDescriptionProps = {
  workload?: Workload;
  health?: H.Health;
  namespace: string;
};

const resourceListStyle = style({
  margin: '0px 0 11px 0',
  $nest: {
    '& > ul > li > span': {
      float: 'left',
      width: '125px',
      fontWeight: 700
    }
  }
});

const iconStyle = style({
  display: 'inline-block',
  verticalAlign: '2px !important'
});

const infoStyle = style({
  margin: '0px 0px 2px 10px',
  verticalAlign: '-5px !important'
});

const healthIconStyle = style({
  marginLeft: '10px',
  verticalAlign: '-1px !important'
});

class WorkloadDescription extends React.Component<WorkloadDescriptionProps> {
  render() {
    const workload = this.props.workload;
    const apps: string[] = [];
    const services: string[] = [];

    if (workload) {
      if (workload.labels[serverConfig.istioLabels.appLabelName]) {
        apps.push(workload.labels[serverConfig.istioLabels.appLabelName]);
      }
      workload.services.forEach(s => services.push(s.name));
    }

    const isTemplateLabels =
      workload &&
      ['Deployment', 'ReplicaSet', 'ReplicationController', 'DeploymentConfig', 'StatefulSet'].indexOf(workload.type) >=
        0;
    const runtimes = (workload?.runtimes || []).map(r => r.name).filter(name => name !== '');

    const workloadProperties = workload ? (
      <>
        <div key="properties-list" className={resourceListStyle}>
          <ul style={{ listStyleType: 'none' }}>
            {workload.istioInjectionAnnotation !== undefined && (
              <li>
                <span>Istio Injection</span>
                {String(workload.istioInjectionAnnotation)}
              </li>
            )}
            <li>
              <span>Type</span>
              {workload.type ? workload.type : 'N/A'}
            </li>
            <li>
              <span>Created</span>
              <div style={{ display: 'inline-block' }}>
                <LocalTime time={workload.createdAt} />
              </div>
            </li>
            <li>
              <span>Version</span>
              {workload.resourceVersion}
            </li>
            {workload.additionalDetails.map((additionalItem, idx) => {
              return (
                <li key={'additional-details-' + idx} id={'additional-details-' + idx}>
                  <span>{additionalItem.title}</span>
                  {additionalItem.icon && renderAPILogo(additionalItem.icon, undefined, idx)}
                  <TextOrLink text={additionalItem.value} urlTruncate={64} />
                </li>
              );
            })}
            {runtimes.length > 0 && (
              <li id="runtimes">
                <span>Runtimes</span>
                <div style={{ display: 'inline-block' }}>
                  {runtimes
                    .map((rt, idx) => renderRuntimeLogo(rt, idx))
                    .reduce(
                      (list: JSX.Element[], elem) =>
                        list.length > 0 ? [...list, <span key="sep"> | </span>, elem] : [elem],
                      []
                    )}
                </div>
              </li>
            )}
          </ul>
        </div>
      </>
    ) : undefined;

    return workload ? (
      <Card id={'WorkloadDescriptionCard'}>
        <CardHeader>
          <Title headingLevel="h5" size="lg">
            <div key="service-icon" className={iconStyle}>
              <PFBadge badge={PFBadges.Workload} position={TooltipPosition.top} />
            </div>
            {this.props.workload ? this.props.workload.name : 'Workload'}
            {workloadProperties ? (
              <Tooltip
                position={TooltipPosition.right}
                content={<div style={{ textAlign: 'left' }}>{workloadProperties}</div>}
              >
                <KialiIcon.Info className={infoStyle} />
              </Tooltip>
            ) : undefined}
            <span className={healthIconStyle}>
              <HealthIndicator id={workload.name} health={this.props.health} />
            </span>
            {this.props.workload && !this.props.workload.istioSidecar && (
              <MissingSidecar
                namespace={this.props.namespace}
                tooltip={true}
                style={{ marginLeft: '10px' }}
                text={''}
              />
            )}
            {this.props.workload && (!this.props.workload.appLabel || !this.props.workload.versionLabel) && (
              <MissingLabel
                missingApp={!this.props.workload.appLabel}
                missingVersion={!this.props.workload.versionLabel}
                style={{ marginLeft: '10px' }}
                tooltip={true}
              />
            )}
          </Title>
        </CardHeader>
        <CardBody>
          {workload.labels && (
            <Labels
              labels={workload.labels}
              tooltipMessage={isTemplateLabels ? 'Labels defined on the Workload template' : undefined}
            />
          )}
          <DetailDescription
            namespace={this.props.namespace}
            apps={apps}
            services={services}
            health={this.props.health}
          />
        </CardBody>
      </Card>
    ) : (
      'Loading'
    );
  }
}

export default WorkloadDescription;
