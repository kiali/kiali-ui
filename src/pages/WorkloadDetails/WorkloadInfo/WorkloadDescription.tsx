import * as React from 'react';
import { Workload } from '../../../types/Workload';
import LocalTime from '../../../components/Time/LocalTime';
import { DisplayMode, HealthIndicator } from '../../../components/Health/HealthIndicator';
import { WorkloadHealth } from '../../../types/Health';
import Labels from '../../../components/Label/Labels';
import {
  Card,
  CardBody,
  Grid,
  GridItem,
  PopoverPosition,
  Stack,
  StackItem,
  Text,
  TextVariants,
  Title
} from '@patternfly/react-core';
import { TextOrLink } from 'components/TextOrLink';
import { renderRuntimeLogo, renderAPILogo } from 'components/Logo/Logos';
import GraphDataSource from '../../../services/GraphDataSource';
import MiniGraphCard from '../../../components/CytoscapeGraph/MiniGraphCard';
import MissingSidecar from '../../../components/MissingSidecar/MissingSidecar';
import { style } from 'typestyle';

const titleStyle = style({
  margin: '15px 0 11px 0'
});

type WorkloadDescriptionProps = {
  workload: Workload;
  namespace: string;
  istioEnabled: boolean;
  health?: WorkloadHealth;
  miniGraphDataSource: GraphDataSource;
};

class WorkloadDescription extends React.Component<WorkloadDescriptionProps> {
  render() {
    const workload = this.props.workload;
    const isTemplateLabels =
      workload &&
      ['Deployment', 'ReplicaSet', 'ReplicationController', 'DeploymentConfig', 'StatefulSet'].indexOf(workload.type) >=
        0;
    const runtimes = workload.runtimes.map(r => r.name).filter(name => name !== '');
    return workload ? (
      <Grid gutter="md">
        <GridItem span={4}>
          <Card style={{ height: '100%' }}>
            <CardBody>
              <Title headingLevel="h3" size="2xl">
                {' '}
                Workload Properties{' '}
              </Title>
              <Stack gutter="md" style={{ marginTop: '10px' }}>
                <StackItem id="name">
                  <Title headingLevel="h6" size="md">
                    {' '}
                    Name{' '}
                  </Title>
                  {workload.name}
                  {!this.props.istioEnabled && (
                    <span style={{ marginLeft: '10px' }}>
                      <MissingSidecar namespace={this.props.namespace} />
                    </span>
                  )}
                </StackItem>
                <StackItem id="labels">
                  <Title headingLevel="h6" size="md">
                    {' '}
                    {isTemplateLabels ? 'Template Labels' : 'Labels'}{' '}
                  </Title>
                  <Labels labels={workload.labels || {}} />
                </StackItem>
                <StackItem id="type">
                  <Title headingLevel="h6" size="md">
                    {' '}
                    Type{' '}
                  </Title>
                  {workload.type ? workload.type : 'N/A'}
                </StackItem>
                <StackItem id="created-at">
                  <Title headingLevel="h6" size="md">
                    {' '}
                    Created at{' '}
                  </Title>
                  <LocalTime time={workload.createdAt} />
                </StackItem>
                <StackItem id="resource-version">
                  <Title headingLevel="h6" size="md">
                    {' '}
                    Resource Version{' '}
                  </Title>
                  {workload.resourceVersion}
                </StackItem>
                {workload.additionalDetails.map((additionalItem, idx) => {
                  return (
                    <StackItem key={'additional-details-' + idx} id={'additional-details-' + idx}>
                      <Title headingLevel="h6" size="md">
                        {' '}
                        {additionalItem.title}{' '}
                      </Title>
                      {additionalItem.icon && renderAPILogo(additionalItem.icon, undefined, idx)}
                      <TextOrLink text={additionalItem.value} urlTruncate={64} />
                    </StackItem>
                  );
                })}
                {runtimes.length > 0 && (
                  <StackItem id="runtimes">
                    <Text component={TextVariants.h3}> Runtimes</Text>
                    {runtimes
                      .map((rt, idx) => renderRuntimeLogo(rt, idx))
                      .reduce(
                        (list: JSX.Element[], elem) =>
                          list.length > 0 ? [...list, <span key="sep"> | </span>, elem] : [elem],
                        []
                      )}
                  </StackItem>
                )}
              </Stack>
            </CardBody>
          </Card>
        </GridItem>
        <GridItem span={4}>
          <MiniGraphCard dataSource={this.props.miniGraphDataSource} />
        </GridItem>
        <GridItem span={4}>
          <Card style={{ height: '100%' }}>
            <CardBody>
              <Title headingLevel="h3" size="2xl">
                {' '}
                Health Overview{' '}
              </Title>
              <Stack gutter="md" style={{ marginTop: '10px' }}>
                <StackItem id="health" className={'stack_service_details'}>
                  <Title headingLevel="h3" size="lg" className={titleStyle}>
                    Overall Health
                  </Title>
                  <HealthIndicator
                    id={workload.name}
                    health={this.props.health}
                    mode={DisplayMode.LARGE}
                    tooltipPlacement={PopoverPosition.left}
                  />
                </StackItem>
              </Stack>
            </CardBody>
          </Card>
        </GridItem>
      </Grid>
    ) : (
      'Loading'
    );
  }
}

export default WorkloadDescription;
