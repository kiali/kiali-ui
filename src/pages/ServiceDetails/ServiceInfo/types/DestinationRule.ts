import { DestinationRuleSpec, K8sMetadata } from '../../../../types/IstioObjects';
import Subset from './Subset';

export default class DestinationRule {
  metadata: K8sMetadata;
  spec: DestinationRuleSpec;
  unformattedField: string;

  constructor(metadata: K8sMetadata, spec: DestinationRuleSpec) {
    this.metadata = metadata;
    this.spec = spec;
  }

  isValid() {
    return this.hasValidName() && this.hasSpecData() && this.hasValidSubsets() && this.hasValidHost();
  }

  hasValidName() {
    if (typeof this.metadata.name !== 'string') {
      this.unformattedField = 'Name';
      return false;
    }

    return true;
  }

  hasSpecData() {
    return this.spec !== null;
  }

  hasValidSubsets() {
    if (!this.spec.subsets) {
      return true;
    }

    let valid = this.spec.subsets instanceof Array;
    valid =
      valid &&
      this.spec.subsets.every((subset, i, ary) =>
        new Subset(subset.name, subset.labels, subset.trafficPolicy).isValid()
      );

    if (!valid) {
      this.unformattedField = 'Subsets';
    }

    return valid;
  }

  hasValidHost() {
    if (typeof this.spec.host !== 'string') {
      this.unformattedField = 'Host';
      return false;
    }

    return true;
  }

  formatValidation() {
    if (!this.isValid()) {
      return {
        message: 'This destination rule has format problems in field ' + this.unformattedField,
        severity: 'error',
        path: ''
      };
    }

    return null;
  }
}
