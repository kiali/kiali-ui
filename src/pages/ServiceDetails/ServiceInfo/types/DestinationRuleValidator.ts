import { DestinationRule } from '../../../../types/IstioObjects';
import SubsetValidator from './SubsetValidator';

export default class DestinationRuleValidator {
  destinationRule: DestinationRule;
  unformattedField: string;

  constructor(destinationRule: DestinationRule) {
    this.destinationRule = destinationRule;
    this.unformattedField = 'none';
  }

  isValid() {
    return this.hasValidName() && this.hasSpecData() && this.hasValidSubsets() && this.hasValidHost();
  }

  hasValidName() {
    if (typeof this.destinationRule.metadata.name !== 'string') {
      this.unformattedField = 'Name';
      return false;
    }

    return true;
  }

  hasSpecData() {
    return this.destinationRule.spec !== null;
  }

  hasValidSubsets() {
    if (!this.destinationRule.spec.subsets) {
      return true;
    }

    let valid = this.destinationRule.spec.subsets instanceof Array;
    valid =
      valid && this.destinationRule.spec.subsets.every((subset, _i, _ary) => new SubsetValidator(subset).isValid());

    if (!valid) {
      this.unformattedField = 'Subsets';
    }

    return valid;
  }

  hasValidHost() {
    if (typeof this.destinationRule.spec.host !== 'string') {
      this.unformattedField = 'Host';
      return false;
    }

    return true;
  }

  formatValidation() {
    if (!this.isValid()) {
      return {
        message: 'This destination rule has format problems in field ' + this.unformattedField,
        severity: 'error',
        path: ''
      };
    }

    return null;
  }
}
