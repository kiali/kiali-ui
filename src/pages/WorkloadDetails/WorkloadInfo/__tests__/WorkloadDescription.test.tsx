import * as React from 'react';
import { shallow } from 'enzyme';
import WorkloadDescription from '../WorkloadDescription';
import { Workload } from '../../../../types/Workload';
import { shallowToJson } from 'enzyme-to-json';

const workload: Workload = {
  name: 'myworkload',
  type: 'Deployment',
  createdAt: '42',
  resourceVersion: '42',
  istioSidecar: false,
  labels: {},
  appLabel: false,
  versionLabel: false,
  replicas: 1,
  availableReplicas: 1,
  pods: [],
  services: [],
  runtimes: [
    {
      name: 'Vert.x',
      dashboardRefs: []
    },
    {
      name: '42',
      dashboardRefs: []
    }
  ]
};

describe('WorkloadDescription', () => {
  it('should render with runtimes', () => {
    const wrapper = shallow(
      <WorkloadDescription workload={workload} namespace={'my-namespace'} istioEnabled={false} />
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
