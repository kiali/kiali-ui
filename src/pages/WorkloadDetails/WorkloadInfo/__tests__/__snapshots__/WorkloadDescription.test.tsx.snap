// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkloadDescription should render with runtimes 1`] = `
<div
  className="card-pf"
>
  <div
    className="card-pf-body"
  >
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        componentClass="div"
        lg={6}
        md={6}
        sm={8}
        xs={12}
      >
        <div
          id="labels"
        >
          <div
            className="progress-description"
          >
            <strong>
              Template Labels
            </strong>
          </div>
          <div
            className="label-collection"
          >
            <Labels
              labels={Object {}}
            />
          </div>
        </div>
        <div>
          <strong>
            Type
          </strong>
           
          Deployment
        </div>
        <div>
          <strong>
            Created at
          </strong>
           
          <LocalTime
            time="42"
          />
        </div>
        <div>
          <strong>
            Resource Version
          </strong>
           
          42
        </div>
        <div>
          <br />
          <img
            alt="Vert.x"
            key="logo-0"
            src="vertx-logo.png"
            title="Vert.x"
          />
          <span
            key="sep"
          >
             | 
          </span>
          <span
            key="runtime-1"
          >
            42
          </span>
        </div>
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        lg={3}
        md={3}
        sm={4}
        xs={12}
      />
      <Col
        bsClass="col"
        componentClass="div"
        lg={3}
        md={3}
        sm={4}
        xs={12}
      >
        <div
          className="progress-description"
        >
          <strong>
            Health
          </strong>
        </div>
        <HealthIndicator
          id="myworkload"
          mode={0}
          tooltipPlacement="left"
        />
      </Col>
    </Row>
  </div>
</div>
`;
